common/pipeline_storage/src/test/scala/uk/ac/wellcome/pipeline_storage/IndexerTest.scala

```diff
@@ -0,0 +1,164 @@
+    }
+  }
+
+  it("replaces a document with the same version") {
+    val document = SampleDocument(3, createCanonicalId, "document")
+    val updatedDocument = document.copy(
+      title = "A different title"
+    )
+
+    withIndexAndIndexer[SampleDocument, Any]() {
+      case (index, indexer) =>
+        val future = for {
+          _ <- indexer.index(Seq(document))
+          result <- indexer.index(Seq(updatedDocument))
+        } yield result
+
+        whenReady(future) { result =>
+          result.right.get should contain(updatedDocument)
+          assertElasticsearchEventuallyHas(index = index, updatedDocument)
+        }
+    }
+  }
+
+  it("inserts a list of documents into elasticsearch and returns them") {
+    val documents =
+      (1 to 5).map(i => SampleDocument(1, createCanonicalId, f"document $i"))
+
+    withIndexAndIndexer[SampleDocument, Any]() {
+      case (index, indexer) =>
+        val future = indexer.index(documents)
+
+        whenReady(future) { successfullyInserted =>
+          assertElasticsearchEventuallyHas(index = index, documents: _*)
+          successfullyInserted.right.get should contain theSameElementsAs documents
+        }
+    }
+  }
+
+  it("returns a list of documents that weren't indexed correctly") {
+    val validDocuments =
+      (1 to 5).map(i => SampleDocument(1, createCanonicalId, f"document $i"))
+    val notMatchingMappingDocuments = (1 to 3).map(
+      i =>
+        SampleDocument(
+          1,
+          createCanonicalId,
+          f"document $i",
+          SampleDocumentData(Some("blah bluh blih"))))
+    val documents = validDocuments ++ notMatchingMappingDocuments
+
+    withIndexAndIndexer[SampleDocument, Any](
+      config = StrictWithNoDataIndexConfig) {
+      case (index, indexer) =>
+        val future = indexer.index(
+          documents = documents
+        )
+
+        whenReady(future) { result =>
+          assertElasticsearchEventuallyHas(index = index, validDocuments: _*)
+          assertElasticsearchNeverHas(
+            index = index,
+            notMatchingMappingDocuments: _*)
+          result.left.get should contain only (notMatchingMappingDocuments: _*)
+        }
+    }
+  }
+
+  def withIndexAndIndexer[T, R](config: IndexConfig = NoStrictMapping)(
+    testWith: TestWith[(Index, Indexer[T]), R])(implicit i: Indexable[T],
+                                                c: CanonicalId[T],
+                                                v: Version[T]) = {
+    withLocalElasticsearchIndex(config) { index =>
+      withIndexer[T, R](index) { indexer =>
+        testWith((index, indexer))
+      }
+    }
+  }
+
+  object StrictWithNoDataIndexConfig extends IndexConfig {
+    import com.sksamuel.elastic4s.ElasticDsl._
+
+    val analysis = WorksAnalysis()
+
+    val title = textField("title")
+    val data = objectField("data")
+
+    val mapping = properties(Seq(title, canonicalId, version, data))
+      .dynamic(DynamicMapping.Strict)
+  }
+}
```

Comment:
Assuming we will be creating an API over the storage layer, and be able to write in memory tests?

Reply:
Yep :+1: